File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
def solution(N, number):
    # DP[값]이 아니라 DP[N 사용 개수]
    dp = [set() for _ in range(9)]
    
    for i in range(1, 9): # 최솟값이 8
        # N, NN, NNN, ... 
        dp[i].add(int(str(N)*i))
        
        # j개 사용한 결과 + (i-j)개 사용한 결과
        for j in range(1, i):
            for a in dp[j]:
                for b in dp[i-j]:
                    dp[i].add(a+b)
                    dp[i].add(a-b)
                    dp[i].add(a*b)
                    if b != 0:
                        dp[i].add(a // b)
    
        if number in dp[i]:
            return i
    return -1

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.cpp
#include <string>
#include <vector>
#include <set>

using namespace std;

int solution(int N, int number) {
    vector<set<int> > dp(9);
    
    for (int i = 1; i < 9; i++) { // i번 사용한 수
        string str = "";
        for (int j = 0; j < i; j++) str += to_string(N);
        dp[i].insert(stoi(str));
        for (int j = 0; j < i/2 + 1; j++) {
            for (auto op1 : dp[j]) {
                for (auto op2 : dp[i - j]) {
                    dp[i].insert(op1 + op2);
                    dp[i].insert(op1 - op2);
                    dp[i].insert(op2 - op1);
                    dp[i].insert(op1 * op2);
                    if (op2 != 0) dp[i].insert(op1 / op2);
                    if (op1 != 0) dp[i].insert(op2 / op1);
                }
            }
        }
        
        if (dp[i].find(number) != dp[i].end()) return i;
    }
    
    return -1;
}